enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/replication/tx/CommittingTransactionsRegistry.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.coreedge.raft.replication.tx;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Future;
+
+import org.neo4j.coreedge.raft.replication.session.LocalOperationId;
+
+public class CommittingTransactionsRegistry implements CommittingTransactions
+{
+    private final Map<LocalOperationId, FutureTxId> outstanding = new ConcurrentHashMap<>();
+
+    @Override
+    public Future<Long> register( LocalOperationId localOperationId )
+    {
+        final FutureTxId future = new FutureTxId( localOperationId );
+        outstanding.put( localOperationId, future );
+        return future;
+    }
+
+    @Override
+    public CompletableFuture<Long> retrieve( LocalOperationId localOperationId )
+    {
+        return outstanding.remove( localOperationId );
+    }
+
+    private class FutureTxId extends CompletableFuture<Long>
```

Comment:
I think it makes sense to make this a top-level class and return that type from register(..). Also don't override cancel but rather add an explicit destroy function or some twr semantics.